core/src/main/java/org/apache/iceberg/BaseUpdateSnapshotRefs.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.Map;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+public class BaseUpdateSnapshotRefs implements UpdateSnapshotRefs {
+
+  private final TableOperations ops;
+  private final TableMetadata base;
+  private final Map<String, SnapshotRef> refs;
+
+  BaseUpdateSnapshotRefs(TableOperations ops) {
+    this.ops = ops;
+    this.base = ops.current();
+    this.refs = Maps.newHashMap(base.refs());
+  }
+
+  @Override
+  public UpdateSnapshotRefs tag(String name, long snapshotId) {
+    Preconditions.checkArgument(name != null, "Tag name must not be null");
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null, "Cannot find snapshot with ID: %s", snapshotId);
+    Preconditions.checkArgument(!refs.containsKey(name), "Cannot tag snapshot, ref already exists: %s", name);
```

Comment:
Should there be a `replaceTag` method for this case? Otherwise, I'd have to remove a tag and re-add it, but I wouldn't really know whether the tag already existed or if it was a tag and not a branch.

Reply:
yeah the current API tries to have just `create`, `remove` and `rename` and ref property update APIs, because it's a bit unclear to me if we need a replace given that there is no such operation in schema and partition spec update.

The action `replace` indicates the tag already exists, and I imagine it should throw exception if the tag does not exist. In that case, user still need to first check existence and then decide to create or replace, and it has little difference from just updating properties of the ref.

If we allow `replace` to mean `put` that either create or update, then my concern is that it might cause unintented use that people might override a tag without knowing it already exists. The current API is basically trying to force a check on user side because there is not really a compare-and-swap of the specific ref that we could enforce as of today on server side and we have to replace the entire metadata file.

any thoughts?

Reply:
I need to think more about what we want in this API vs in the other APIs. Probably worth thinking through the TableMetadata changes and API as well. It could be that we want to have this style API there instead of here and just keep this to high level operations.